lib/TypeNames/DependencyGraph.cpp

```diff
@@ -216,156 +194,171 @@ DependencyGraph::Builder::addArtificialNodes(const model::RawFunctionDefinition
+
+  const auto &[DependentDeclNode, DependentDefNode] = Dependent;
+  const model::TypeDefinition *DependentDefinition = DependentDeclNode->T;
+
+  revng_assert(DependentDeclNode);
+  revng_assert(DependentDefinition);
+
+  TypeDependencyNode *DependentNode = nullptr;
+  bool
+    HasForwardDeclaration = hasSeparateForwardDeclaration(*DependentDefinition)
+                            or DependentDeclNode->isArtificial();
+  if (HasForwardDeclaration) {
+    revng_assert(DependentDefNode);
+    DependentNode = DependentDefNode;
+  } else {
+    revng_assert(not DependentDefNode);
+    DependentNode = DependentDeclNode;
+  }
+  revng_assert(DependentNode);
+
+  AssociatedNodes NodesDependedOn = Graph->TypeToNodes.at(DefinitionDependedOn);
+  revng_assert(NodesDependedOn.Declaration);
+  revng_assert(not NodesDependedOn.Declaration->isArtificial());
+  revng_assert(not NodesDependedOn.Definition
+               or not NodesDependedOn.Definition->isArtificial());
+
+  // If LastArray is true, the node depended on is always the node
+  // representing the full definition of the type, which in some cases might
+  // be the Declaration node (e.g. when the type depended on doesn't have a
+  // separate Definition and Declaration node, but only a Declaration, such as
+  // for TypedefDefinitions and function type definitions).
+  if (LastArray) {
+    TypeDependencyNode *NodeDependedOn = NodesDependedOn.Definition ?
+                                           NodesDependedOn.Definition :
+                                           NodesDependedOn.Declaration;
+    revng_assert(NodeDependedOn);
+    addAndLogSuccessor(DependentNode, NodeDependedOn);
+    return;
   }
 
-  if (PointerIsBetweenTypes) {
-    // Otherwise, if we've found at least a pointer, we only depend on the name
-    // of the pointee.
-    return Graph->TypeToNodes.at(EdgeTarget).Declaration;
+  // If FoundPointer is true, and LastArray is false, the node depended on is
+  // always the Declaration node, because we don't need the full definition.
+  if (FoundPointer) {
+    addAndLogSuccessor(DependentNode, NodesDependedOn.Declaration);
+    return;
   }
 
-  // In all the other cases we depend on the type with the kind indicated by K.
-  if constexpr (K == TypeNode::Kind::Declaration)
-    return Graph->TypeToNodes.at(EdgeTarget).Declaration;
-  else if constexpr (K == TypeNode::Kind::Definition)
-    return Graph->TypeToNodes.at(EdgeTarget).Definition;
-  else
-    static_assert(value_always_false_v<K>);
+  // Otherwise we fall back in the baseline case.
 
-  return nullptr;
+  // The DependentNode always depends on the the Declaration node of DependedOn.
+  addAndLogSuccessor(DependentNode, NodesDependedOn.Declaration);
+
+  // Then if both Dependent and DependedOn has a separate forward declaration we
+  // add a dependency from DependentNode to the Definition of the DependedOn.
+  if (HasForwardDeclaration
+      and hasSeparateForwardDeclaration(*DefinitionDependedOn)) {
```

Comment:
This case is extra confusing for me, isn't `hasSeparateForwardDeclaration` `true` for structs? I'm probably missing something, but why'd we want to depend on a struct body from a _declaration_ of another struct?

Reply:
Ah, fuck, I just realized that I just misread the 252-258 `if`.

For some reason, in my mind, the dependencies were added for both half-nodes at once. Yeah, that thought line doesn't make any sense!

Sorry for the noise, close this thread please.